modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/MetaStorageManagerTest.java

```diff
@@ -0,0 +1,297 @@
+
+        return nodes.stream()
+            .map(node -> {
+                VaultManager vaultManager = mock(VaultManager.class);
+
+                when(vaultManager.name()).thenReturn(CompletableFuture.completedFuture(node.name()));
+
+                var metaStorageManager = new MetaStorageManager(
+                    vaultManager, clusterService, raftManager, msgFactory, config
+                );
+
+                return new Node(node, metaStorageManager);
+            })
+            .collect(Collectors.toUnmodifiableList());
+    }
+
+    /**
+     * Creates a {@code ClusterNode} on a random port.
+     */
+    private static ClusterNode createNode(String name) {
+        int port = ThreadLocalRandom.current().nextInt(0, 10000);
+
+        return new ClusterNode(name, name, new NetworkAddress("localhost", port));
+    }
+
+    /**
+     * Tests the startup of a single node that hosts the Meta Storage.
+     * It should simply block until the "init" command arrives.
+     */
+    @Test
+    void testSingleNodeHostStart() {
+        Node node = createCluster("foobar").get(0);
+
+        node.manager.start();
+
+        verify(messagingService, never()).send(any(NetworkAddress.class), any(), anyString());
+
+        CompletableFuture<Boolean> hasMetastorage = supplyAsync(node.manager::hasMetastorageLocally);
+
+        node.manager.init(List.of(node.clusterNode.name()));
+
+        assertThat(hasMetastorage, willBe(equalTo(true)));
+    }
+
+    /**
+     * Tests the startup of a single node that does not host the Meta Storage.
+     * It should simply block until the "init" command arrives.
+     */
+    @Test
+    void testSingleNodeNotHostStart() {
+        Node node = createCluster("foobar").get(0);
+
+        node.manager.start();
+
+        verify(messagingService, never()).send(any(NetworkAddress.class), any(), anyString());
+
+        CompletableFuture<Boolean> hasMetastorage = supplyAsync(node.manager::hasMetastorageLocally);
+
+        node.manager.init(List.of("barbaz"));
+
+        assertThat(hasMetastorage, willBe(equalTo(false)));
+    }
+
+    /**
+     * Tests a scenario when two nodes start and receive a broadcast "init" command.
+     */
+    @Test
+    void testZombieStateStart() {
+        NetworkMessage response = msgFactory.metastorageNotReadyResponse().build();
+
+        when(messagingService.invoke(any(ClusterNode.class), any(), anyLong()))
+            .thenReturn(CompletableFuture.completedFuture(response));
+
+        List<Node> nodes = createCluster("foo", "bar");
+
+        nodes.forEach(node -> node.manager.start());
+
+        CompletableFuture<Boolean> hasMetastorageNode0 = supplyAsync(nodes.get(0).manager::hasMetastorageLocally);
+        CompletableFuture<Boolean> hasMetastorageNode1 = supplyAsync(nodes.get(1).manager::hasMetastorageLocally);
+
+        var metaStorageNodes = List.of(nodes.get(0).clusterNode.name());
+
+        nodes.forEach(node -> node.manager.init(metaStorageNodes));
```

Comment:
What's the polling interval in tests? Still 1 second?

Reply:
please see the comment on line 106

Reply:
Cool